File: top_interview_questions/1_easy/4_trees/binary_tree_lever_order_traversal/solution.py
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def __init__(self):
        self.levels = []

    def levelOrder(self, root: TreeNode, level=0) -> List[List[int]]:
        if not root:
            return

        if len(self.levels) < level + 1:
            self.levels.append([])

        self.levels[level].append(root.val)

        self.levelOrder(root.left, level + 1)
        self.levelOrder(root.right, level + 1)

        return self.levels

File: top_interview_questions/1_easy/4_trees/binary_tree_lever_order_traversal/solution.go
package main

import "fmt"

func main() {
	tree := &TreeNode{
		Val: 2,
		Left: &TreeNode{
			Val: 1,
		},
		Right: &TreeNode{
			Val: 3,
		},
	}
	fmt.Println(levelOrder(tree))
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func levelOrder(root *TreeNode) [][]int {
	depth := maxDepth(root)
	levels := make([][]int, depth)
	fmt.Println(levels)
	helper(root, 0, levels)

	return levels
}

func helper(root *TreeNode, level int, levels [][]int) {
	if root == nil || level == len(levels) {
		return
	}

	levels[level] = append(levels[level], root.Val)
	helper(root.Left, level+1, levels)
	helper(root.Right, level+1, levels)
}

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	leftMaxDepth := maxDepth(root.Left)
	rightMaxDepth := maxDepth(root.Right)

	if leftMaxDepth > rightMaxDepth {
		return leftMaxDepth + 1
	}

	return rightMaxDepth + 1
}
